### PR TITLE
W3ddisplay

### PR DESCRIPTION
**Pull request by duncanspumpkin**
~~There are a lot of functions here that can't be done yet but this at least gives us a start.~~ 

All functions that can be done atm I think have now been implemented. The screenshot function looks like it could be done but I do not feel comfortable doing it atm without further exposure to how all the filesystem and dx things interact.

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
~~Hmm a few little bugs right now with drawing boxes. Will try investigate. I've also found a few more functions I'd like to implement.~~ All fixed

File: src/w3d/renderer/assetmgr.h [hunk 1]
```diff
@@ -35,8 +35,8 @@ class W3DAssetManager
 {
 public:
     virtual ~W3DAssetManager();
-    virtual bool Load_3D_Assets(const char *filename);
     virtual bool Load_3D_Assets(FileClass &asset_file);
```
Comment (tomsons26) [inline src/w3d/renderer/assetmgr.h]: is the order reversed when compiled?
it should be 
char
fileclass
in compiled binary

File: src/platform/w3dengine/client/w3ddisplay.h [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    static W3DAssetManager *Get_AssetManager() { return m_assetManager; }
+    static SceneClass *Get_3DInterfaceScene() { return m_3DInterfaceScene; }
+
+private:
+    bool m_initialized;
+    LightClass *m_myLight[4];
+    Render2DClass *m_2DRender;
+    IRegion2D m_clipRegion;
+    bool m_isClippedEnabled;
+    float m_fpsLimit;
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.h]: m_averageFPS

File: src/platform/w3dengine/client/w3ddisplay.h [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    uint32_t m_displayStrings[16];
+    uint32_t unk1;
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.h]: m_benchmarkDisplayString

File: src/platform/w3dengine/client/w3ddisplay.h [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    virtual int Get_Display_Mode_Count() override;
+    virtual void Get_Display_Mode_Description(int id, int *width, int *height, int *bit_depth) override;
+    virtual void Do_Smart_Asset_Purge_And_Preload(const char *) override;
+    virtual VideoBuffer *Create_VideoBuffer() override;
+    virtual void Set_Clip_Region(IRegion2D *region) override;
+    virtual bool Is_Clipping_Enabled() override { return m_isClippedEnabled; }
+    virtual void Enable_Clipping(bool isEnabled) override { m_isClippedEnabled = isEnabled; }
+    virtual void Set_Time_Of_Day(TimeOfDayType time) override;
+    virtual void Create_Light_Pulse(const Coord3D *, const RGBColor *, float, float, unsigned, unsigned) override;
+    virtual void Draw_Line(int32_t x1, int32_t y1, int32_t x2, int32_t y2, float width, uint32_t colour) override;
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.h]: check if order is right in compiled binary, 8 args first ,7 args after

File: src/platform/w3dengine/client/w3ddisplay.h [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    ~W3DDisplay();
+    virtual void Init() override;
+    virtual void Reset() override;
+    virtual void Draw() override;
+    virtual void Set_Width(uint32_t width) override;
+    virtual void Set_Height(uint32_t height) override;
+    virtual bool Set_Display_Mode(uint32_t width, uint32_t height, uint32_t bits, bool windowed) override;
+    virtual int Get_Display_Mode_Count() override;
+    virtual void Get_Display_Mode_Description(int id, int *width, int *height, int *bit_depth) override;
+    virtual void Do_Smart_Asset_Purge_And_Preload(const char *) override;
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.h:39]: virtual Set_Gamma is missing

File: src/platform/w3dengine/client/w3ddisplay.h [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    static SceneClass *Get_3DInterfaceScene() { return m_3DInterfaceScene; }
+
+private:
+    bool m_initialized;
+    LightClass *m_myLight[4];
+    Render2DClass *m_2DRender;
+    IRegion2D m_clipRegion;
+    bool m_isClippedEnabled;
+    float m_fpsLimit;
+    uint32_t m_displayStrings[16];
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.h]: before this add a if WB unknown double member

File: src/platform/w3dengine/client/w3ddisplay.h (hunk 5 above)
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.h:39]: after Do_Smart_Asset_Purge_And_Preload add a comment that there's a virtual in WB here

File: src/platform/w3dengine/client/w3ddisplay.h [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    virtual void Draw_Remaining_Rect_Clock(
+        int32_t x, int32_t y, int32_t width, int32_t height, int percentage, uint32_t colour) override;
+    virtual void Draw_Image(
+        Image *image, int32_t left, int32_t top, int32_t right, int32_t bottom, uint32_t colour, DrawImageMode mode)
+        override;
+    virtual void Draw_VideoBuffer(VideoBuffer *vb, int32_t x1, int32_t y1, int32_t x2, int32_t y2) override;
+    virtual void Set_Shroud_Level(int, int, CellShroudStatus) override;
+    virtual void Clear_Shroud() override{};
+    virtual void Set_Border_Shroud_Level(uint8_t level) override;
+    virtual void Preload_Model_Assets(Utf8String model) override;
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.h:64]: after Set_Border_Shroud_Level add a comment that there's a virtual in WB here

File: src/w3d/renderer/assetmgr.h (hunk 1 above)
Comment (duncanspumpkin) [reply src/w3d/renderer/assetmgr.h]: From my investigation in the asset manager pr this was back to front in thyme. The first virtual was dtor followed by FileClass and the third was const char*.

File: src/platform/w3dengine/client/w3ddisplay.h (hunk 4 above)
Comment (duncanspumpkin) [reply src/platform/w3dengine/client/w3ddisplay.h]: Okay I've swapped that now in Display and W3DDisplay

File: src/platform/w3dengine/client/w3ddisplay.cpp [hunk 8]
```diff
@@ -0,0 +1,950 @@
...
+    // TODO: Requires RTS3DInterfaceScene, RTS2DScene, RTS3DScene, W3DAssetManager, ...
+#ifdef GAME_DLL
+    Call_Method<void, W3DDisplay>(0x0073CA80, this);
+#endif
+}
+
+// 0x0073D030
+void W3DDisplay::Reset()
+{
+    auto *iter = m_3DScene->Create_Iterator(false);
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.cpp]: Needs to call Display::Reset()

File: src/platform/w3dengine/client/w3ddisplay.cpp [hunk 9]
```diff
@@ -0,0 +1,950 @@
...
+        m_3DScene->Remove_Render_Object(renderObj);
+        renderObj->Release_Ref();
+    }
+    m_3DScene->Destroy_Iterator(iter);
+
+    m_isClippedEnabled = false;
+
+    m_assetManager->Release_Unused_Assets();
+
+    g_theWriteableGlobalData->m_drawSkyBox = false;
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.cpp]: check if g_theWriteableGlobalData isn't null missing

File: src/platform/w3dengine/client/w3ddisplay.cpp [hunk 10]
```diff
@@ -0,0 +1,950 @@
...
+    RectClass rect = { { 0.0f, 0.0f }, { static_cast<float>(Get_Width()), static_cast<float>(Get_Height()) } };
+    m_2DRender->Set_Coordinate_Range(rect);
+}
+
+// 0x0073C840
+bool W3DDisplay::Set_Display_Mode(uint32_t width, uint32_t height, uint32_t bits, bool windowed)
+{
+    auto success = DX8Wrapper::Set_Device_Resolution(width, height, bits, windowed, true);
+
+    if (success) {
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.cpp:161]: if DX8Wrapper::Set_Device_Resolution(width, height, bits, windowed, true) == W3D_ERROR_OK

File: src/platform/w3dengine/client/w3ddisplay.cpp [hunk 11]
```diff
@@ -0,0 +1,950 @@
...
+    }
+    return valid_resolutions;
+}
+
+// 0x0073C650
+void W3DDisplay::Get_Display_Mode_Description(int id, int *width, int *height, int *bit_depth)
+{
+    auto &desc = DX8Wrapper::Get_Render_Device_Desc(0);
+    auto &resolutions = desc.Get_Resolution_Array();
+    if (resolutions.Count() <= 0) {
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.cpp]: no need, the compiler makes these for loops

File: src/platform/w3dengine/client/w3ddisplay.cpp [hunk 12]
```diff
@@ -0,0 +1,950 @@
...
+
+#ifndef GAME_DLL
+W3DAssetManager *W3DDisplay::m_assetManager;
+SceneClass *W3DDisplay::m_3DScene; // TODO: Actual type is RTS2DScene
+SceneClass *W3DDisplay::m_2DScene; // TODO: Actual type is RTS2DScene
+SceneClass *W3DDisplay::m_3DInterfaceScene; // TODO: Actual type is RTS3DInterfaceScene
+#endif
+
+// 0x0073C3C0
+W3DDisplay::W3DDisplay() : m_initialized(false), m_2DRender(0), m_clipRegion({ { 0, 0 }, { 0, 0 } })
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.cpp]: missing m_isClippedEnabled

File: src/platform/w3dengine/client/w3ddisplay.cpp [hunk 13]
```diff
@@ -0,0 +1,950 @@
...
+#ifdef GAME_DLL
+    Call_Method<void, W3DDisplay, uint8_t>(0x00740CB0, this, level);
+#endif
+}
+
+// 0x007411F0
+void W3DDisplay::Preload_Model_Assets(Utf8String model)
+{
+    if (m_assetManager != nullptr) {
+        Utf8String filename = model;
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.cpp]: Utf8String filename;
        filename.Format("%s.w3d", model.str());
        m_assetManager->Load_3D_Assets(filename);

File: src/platform/w3dengine/client/w3ddisplay.cpp [hunk 14]
```diff
@@ -0,0 +1,950 @@
...
+    if (m_windowed == true) {
+        return;
+    }
+
+    DX8Wrapper::Set_Gamma(gamma, bright, contrast, calibrate, false);
+}
+
+// 0x00741360 Unsure if used
+void W3DDisplay::Do_Smart_Asset_Purge_And_Preload(const char *asset)
+{
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.cpp:260]: needs m_assetManager != nullptr

File: src/platform/w3dengine/client/w3ddisplay.h [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+    Render2DClass *m_2DRender;
+    IRegion2D m_clipRegion;
+    bool m_isClippedEnabled;
+    float m_averageFps;
+    DisplayString *m_displayStrings[16];
+    DisplayString *m_benchmarkDisplayString;
+    uint32_t m_nativeDebugDisplay;
+
+#ifdef GAME_DLL
+    static W3DAssetManager *&m_assetManager;
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.h]: all should be s_

File: src/platform/w3dengine/client/w3ddisplay.cpp (hunk 10 above)
Comment (duncanspumpkin) [reply src/platform/w3dengine/client/w3ddisplay.cpp:161]: The Set_Device_Resolution function returns a bool are you sure this is correct.

Comment (duncanspumpkin) [reply src/platform/w3dengine/client/w3ddisplay.cpp:161]: I think the issue is this was wrapped by WW3D which modified the return value.

Comment (duncanspumpkin) [reply src/platform/w3dengine/client/w3ddisplay.cpp:161]: Wrapper has been added.

File: src/platform/w3dengine/client/w3ddisplay.cpp [hunk 16]
```diff
@@ -0,0 +1,956 @@
...
+// 0x0073C3C0
+W3DDisplay::W3DDisplay() :
+    m_initialized(false), m_2DRender(0), m_isClippedEnabled(false), m_clipRegion({ { 0, 0 }, { 0, 0 } })
+{
+    s_assetManager = nullptr;
+    s_3DScene = nullptr;
+    s_2DScene = nullptr;
+    s_3DInterfaceScene = nullptr;
+    m_averageFps = g_theWriteableGlobalData->m_framesPerSecondLimit;
+    for (auto &str : m_displayStrings) {
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.cpp:52]: if WB m_unknqword = 0;

File: src/platform/w3dengine/client/w3ddisplay.h [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+    static SceneClass *Get_3DInterfaceScene() { return s_3DInterfaceScene; }
+
+private:
+    bool m_initialized;
+    LightClass *m_myLight[4];
+    Render2DClass *m_2DRender;
+    IRegion2D m_clipRegion;
+    bool m_isClippedEnabled;
+    float m_averageFps;
+    // WB additional unknown double here
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.h]: if WB _QWORD m_unknqword;
this has something to do with performance tracking, it gets the value from QueryPerformanceCounter()

File: src/platform/w3dengine/client/w3ddisplay.h [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+    virtual void Set_Gamma(float gamma, float bright, float contrast, bool calibrate) override;
+    virtual void Do_Smart_Asset_Purge_And_Preload(const char *) override;
+    // WB additional virtual here
+    virtual VideoBuffer *Create_VideoBuffer() override;
+    virtual void Set_Clip_Region(IRegion2D *region) override;
+    virtual bool Is_Clipping_Enabled() override { return m_isClippedEnabled; }
+    virtual void Enable_Clipping(bool isEnabled) override { m_isClippedEnabled = isEnabled; }
+    virtual void Set_Time_Of_Day(TimeOfDayType time) override;
+    virtual void Create_Light_Pulse(const Coord3D *, const RGBColor *, float, float, unsigned, unsigned) override;
+    virtual void Draw_Line(int32_t x1, int32_t y1, int32_t x2, int32_t y2, float width, uint32_t colour) override;
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.h]: for consistency use color, as that's what we have everywhere else already

File: src/platform/w3dengine/client/w3ddisplay.cpp [hunk 19]
```diff
@@ -0,0 +1,956 @@
...
+
+        // Right Bottom Box
+        RectClass right_bottom_rect = { { mid_left, mid_top }, { right, bottom } };
+        m_2DRender->Add_Rect(right_bottom_rect, 0.0f, 0, colour);
+
+        // At 13, clock points at corner of Right Top of box
+        if (percentage < 13) {
+            // Right Top Static Wedge
+            {
+                Vector2 pt0{ mid_left, mid_top };
```
Comment (tomsons26) [inline src/platform/w3dengine/client/w3ddisplay.cpp]: move pt0 out where uv's are since its always the middle

**Comment by tomsons26** [comment]
Is_LetterBox_Fading still seems sketchy to me for some reason, Draw_Image isn't quite exactly the same but should work the same.
I think i done my part, passing on to @OmniBlade now

**Review by OmniBlade — APPROVED** [review]

**Pull request merged**